creusot/src/analysis/not_final_places.rs

```diff
@@ -0,0 +1,279 @@
+    ///
+    /// Two places `p1` and `p2` are conflicting if one is a subplace of the other.
+    conflicting_places: Vec<Vec<PlaceAsSLice<'tcx>>>,
+}
+
+impl<'tcx> NotFinalPlaces<'tcx> {
+    pub fn new(body: &mir::Body<'tcx>) -> Self {
+        struct VisitAllPlaces<'tcx>(HashMap<PlaceAsSLice<'tcx>, PlaceId>);
+        impl<'tcx> Visitor<'tcx> for VisitAllPlaces<'tcx> {
+            fn visit_place(
+                &mut self,
+                place: &Place<'tcx>,
+                _: mir::visit::PlaceContext,
+                _: Location,
+            ) {
+                let mut place_as_slice = PlaceAsSLice::from(place);
+                loop {
+                    let idx = self.0.len();
+                    if let std::collections::hash_map::Entry::Vacant(entry) =
+                        self.0.entry(place_as_slice)
+                    {
+                        entry.insert(idx);
+                    }
+                    if place_as_slice.projection.is_empty() {
+                        break;
+                    }
+                    // Include all "superplace" of the place.
+                    // This is needed, because in `is_final_at`, we might refer to such a
+                    // place, even though it does not appear in the MIR.
+                    place_as_slice.projection =
+                        &place_as_slice.projection[..place_as_slice.projection.len() - 1]
+                }
+            }
+        }
+
+        let mut visitor = VisitAllPlaces(HashMap::new());
+        visitor.visit_body(body);
+        let places = visitor.0;
+        let mut contains_dereference = BitSet::new_empty(places.len());
+        let mut subplaces = vec![Vec::new(); places.len()];
+        let mut conflicting_places = vec![Vec::new(); places.len()];
+        for (place, id) in &places {
+            if place.projection.get(1..).unwrap_or_default().contains(&ProjectionElem::Deref) {
+                contains_dereference.insert(*id);
+            }
+            for (other_place, other_id) in &places {
+                if id == other_id || place.local != other_place.local {
+                    continue;
+                }
+                if other_place.projection.get(..place.projection.len()) == Some(place.projection) {
+                    subplaces[*id].push(*other_place);
+                    conflicting_places[*id].push(*other_place);
+                    conflicting_places[*other_id].push(*place);
+                }
+            }
+        }
+        Self { places, has_indirection: contains_dereference, subplaces, conflicting_places }
+    }
+
+    /// Run the analysis right **after** `location`, and determines if `place` is a final reborrow.
+    pub fn is_final_at(
+        cursor: &mut ResultsCursor<'_, 'tcx, NotFinalPlaces<'tcx>>,
+        place: &Place<'tcx>,
+        location: Location,
+    ) -> bool {
+        let deref_position =
+            match place.projection.iter().position(|p| matches!(p, ProjectionElem::Deref)) {
```

Comment:
You should not consider Deref if the dereferenced type is `Box<T>`

Reply:
This should now be correctly handled, see `NotFinalPlaces::has_indirection`